visualization/motion_planning_tasks/src/task_display.cpp

```diff
@@ -0,0 +1,265 @@
+{
+
+TaskDisplay::TaskDisplay() : Display()
+{
+	robot_description_property_ =
+	      new rviz::StringProperty("Robot Description", "robot_description", "The name of the ROS parameter where the URDF for the robot is loaded",
+	                               this, SLOT(changedRobotDescription()), this);
+
+	task_solution_topic_property_ =
+	      new rviz::RosTopicProperty("Task Solution Topic", "",
+	                                 ros::message_traits::datatype<moveit_task_constructor_msgs::Solution>(),
+	                                 "The topic on which task solutions (moveit_msgs::Solution messages) are received",
+	                                 this, SLOT(changedTaskSolutionTopic()), this);
+
+	trajectory_visual_.reset(new TaskSolutionVisualization(this, this));
+	tasks_property_ =
+	      new rviz::Property("Tasks", QVariant(), "Tasks received on monitored topic", this);
+}
+
+void TaskDisplay::onInitialize()
+{
+	Display::onInitialize();
+	trajectory_visual_->onInitialize(scene_node_, context_);
+}
+
+void TaskDisplay::loadRobotModel()
+{
+	rdf_loader_.reset(new rdf_loader::RDFLoader(robot_description_property_->getStdString()));
+
+	if (!rdf_loader_->getURDF())
+	{
+		this->setStatus(rviz::StatusProperty::Error, "Robot Model",
+		                "Failed to load from parameter " + robot_description_property_->getString());
+		return;
+	}
+	this->setStatus(rviz::StatusProperty::Ok, "Robot Model", "Successfully loaded");
+
+	const srdf::ModelSharedPtr& srdf =
+	      rdf_loader_->getSRDF() ? rdf_loader_->getSRDF() : srdf::ModelSharedPtr(new srdf::Model());
+	robot_model_.reset(new robot_model::RobotModel(rdf_loader_->getURDF(), srdf));
+
+	// Send to child class
+	trajectory_visual_->onRobotModelLoaded(robot_model_);
+	trajectory_visual_->onEnable();
+}
+
+void TaskDisplay::reset()
+{
+	Display::reset();
+	loadRobotModel();
+	trajectory_visual_->reset();
+}
+
+void TaskDisplay::onEnable()
+{
+	Display::onEnable();
+	loadRobotModel();
+
+	// (re)initialize task model
+	updateTaskListModel();
+}
+
+void TaskDisplay::onDisable()
+{
+	Display::onDisable();
+	trajectory_visual_->onDisable();
+
+	// don't monitor topics when disabled
+	task_description_sub.shutdown();
+	task_statistics_sub.shutdown();
+	task_solution_sub.shutdown();
+}
+
+void TaskDisplay::update(float wall_dt, float ros_dt)
+{
+	Display::update(wall_dt, ros_dt);
+	mainloop_jobs_.executeJobs();
+	trajectory_visual_->update(wall_dt, ros_dt);
+}
+
+void TaskDisplay::setName(const QString& name)
+{
+	BoolProperty::setName(name);
+	trajectory_visual_->setName(name);
+}
+
+void TaskDisplay::changedRobotDescription()
+{
+	if (isEnabled())
+		reset();
+	else
+		loadRobotModel();
+}
+
+void TaskDisplay::taskDescriptionCB(const ros::MessageEvent<const moveit_task_constructor_msgs::TaskDescription> &event)
+{
+	const moveit_task_constructor_msgs::TaskDescriptionConstPtr& msg = event.getMessage();
+	const std::string id = event.getPublisherName() + "/" + msg->id;
```

Comment:
This approach fails the moment you replay data, right?

Would it be better to extract the first part of your id from the topic the display subscribes to?

Do we really need this prefix in the id here?

Reply:
The publisher name _is_ the full topic name, at least it was in my test.

Reply:
The inline documentation of `getPublisherName` says

> Returns the name of the node which published this message

To the best of my knowledge that's not the full topic.
If you get the topic name there and plan to use it, you should open an issue in ros_comm and sort out what *should* come out of this function.
I would prefer to avoid using `ros::MessageEvent` in the first place and get the topic name from the display property.